### ISSUE TITLE
Toplevel outputs "type/n" unnecessarily

### ISSUE THREAD

**Issue by xclerc** [ocaml/ocaml#8610]
The following sequence in the toplevel:

```
# type list;;
type list
# let f () = [];;
val f : unit -> 'a list/2 = <fun>
```

seems a bit unexpected, as the "/2" is unnecessary.

(I have no idea whether it is a correct fix, but I noticed
that adding `printing_env := Env.empty;` to the
`Printtyp.Naming_context.reset` function seemingly
fixes the problem.)

**Comment by xclerc** [comment]
cc @Octachron

**Comment by Octachron** [comment]
This can be easily disabled, but this is a conscious choice due to the fact that the type constructor `list` in the type of `f` is not the one currently in scope.
For instance, without this differentiation copying the type displayed by the toplevel and adding it as a constraint to f,
```OCaml
let g: unit -> 'a list = f
```
yields an error.

**Comment by xclerc** [comment]
OK, thanks for the explanation.

**Comment by trefis** [comment]
I think there something @xclerc omitted from his original message, so let me expand on his example:
```
# type list;;
type list
# [ 1 ];;
- : int list = [1]
# let l = [ 1 ];;
val l : int list/2 = [1]
# [ 1 ];;
- : int list/2 = [1]
```

Seems like some state doesn't get updated in "toplevel expression" branch.

**Comment by Octachron** [comment]
Indeed, there is a problem with the end state of `Printtyp.print_items` if the last item is a type. This problematic end state is however overwritten by subsequent calls to `Printtyp.print_items` and is thus only observable with toplevel expressions that call directly `Printtyp.tree_of_type_scheme`.

**Comment by Octachron** [comment]
Fixed by #8613 .
